src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -116,19 +170,18 @@ public AssetProvider CreateAssetProvider(Checksum solutionChecksum, SolutionAsse
 
             try
             {
-                return await ProcessSolutionAsync(inFlightSolution, solutionTask).ConfigureAwait(false);
-            }
-            catch (Exception ex) when (FatalError.ReportAndPropagateUnlessCanceled(ex, cancellationToken, ErrorSeverity.Critical))
-            {
-                // Any non-cancellation exception is bad and needs to be reported.  We will still ensure that we cleanup
-                // below though no matter what happens so that other calls to OOP can properly work.
-                throw ExceptionUtilities.Unreachable();
+                await foreach (var pair in ProcessSolutionAsync(inFlightSolution, solutionTask, cancellationToken).ConfigureAwait(false))
+                    yield return pair;
             }
```

Comment:
cant have a `yield` in a catch (not sure why @jcouv ?), so this `catch` moved to the outer async helper function.

Reply:
here all the work happens betwenen the `AcquireAndIncrement` above and the `finally-Decrement` below.

Reply:
> can't have a yield in a catch

That's by [design](https://github.com/dotnet/roslyn/blob/main/docs/features/async-streams.md): "It is a compile-time error for a yield return statement to appear anywhere in a try statement that contains any catch clauses."
There's a proposal to support it (seems possible if I recall correctly, but would need some investigation to confirm the language rules as `catch` block can evaluate during disposal): https://github.com/dotnet/roslyn/issues/39583